feat: date filters in audit trail

GursheenK opened the pull request:
**Changes**

- Added condition in the Document field to filter out any non-amended docs.
- Added Date Range filters to show the audit trail only within the selected range.

**Screenshot**

<img width="1296" alt="Screenshot 2023-12-03 at 5 32 42 PM" src="https://github.com/frappe/frappe/assets/40693548/8855bcb9-3acd-4e48-bfcf-59ab8a0e47e5">

<br>
<br>

Updated docs: https://frappeframework.com/docs/user/en/audit-trail

GursheenK committed:
fix: only show amended docs in dropdown

GursheenK committed:
feat: start and end date filters

GursheenK committed:
fix: make date filters mandatory

GursheenK committed:
fix: filter document field using date

GursheenK committed:
fix: validate date fields explicitly

GursheenK committed:
fix: filter docs in controller

GursheenK committed:
fix: set date fields in test

GursheenK committed:
Merge branch 'develop' into audit-trail-filters

GursheenK committed:
chore: fix translation

surajshetty3416 reviewed (commented):
As discussed

- [x] Make filters optional
- [x] Provide option to show audit trail from transaction records

GursheenK committed:
fix: make date range filters optional

GursheenK committed:
fix: remove validation for date filters

GursheenK committed:
fix: filter query for non mandatory filters

GursheenK committed:
feat: show audit trail from transaction records

GursheenK committed:
feat: set filters when routed from transaction

GursheenK commented:
@surajshetty3416   @deepeshgarg007 

Have made the date range filters optional and moved them into a collapsible. In order to ensure the document field uses the optional dates correctly, it will unset the selected document if it does not belong to the date range, otherwise keep it as it is. 

Also added an option in the Menu to show the Audit Trail from the Transaction Records. Currently using the previous route for fetching the DocType and Document name since the `route_options` do not get set in the url for a single doctype. Please lmk if there's a better solution to implement this.

surajshetty3416 commented on frappe/core/doctype/audit_trail/audit_trail.js:
```diff
@@ -2,6 +2,17 @@
 // For license information, please see license.txt
 
 frappe.ui.form.on("Audit Trail", {
+	onload(frm) {
+		let prev_route = frappe.get_prev_route();
+		if (prev_route.length > 2 && prev_route[0] == "Form" && prev_route[1] != "Audit Trail") {
+			frm.doc.doctype_name = prev_route[1];
+			frm.doc.document = prev_route[2];
+			frm.doc.start_date = "";
+			frm.doc.end_date = "";
+		}
+		frm.events.get_audit_trail_for_document(frm);
+	},
```
You might have to move this to refresh with some additional checks. Currently, once you load audit trail, then go back and open audit trail from different transaction record, filters are not updated.

Also, it seems like a bug that single form does not load route_options. [created an issue](https://github.com/frappe/frappe/issues/23061)

surajshetty3416 commented on frappe/public/js/frappe/form/toolbar.js:
```diff
@@ -498,6 +498,19 @@ frappe.ui.form.Toolbar = class Toolbar {
+			frappe.model.get_value("DocType", this.frm.doc.doctype, "track_changes")
+		) {
+			this.page.add_menu_item(
+				__("Audit Trail"),
```
Let's call it "View Audit Trail"?

GursheenK committed:
fix: set filters in refresh

GursheenK committed:
fix: change btn label
